### ISSUE TITLE
Mongoose browser does not validate array of schemas

### ISSUE THREAD

**Issue by RamAkunuru** [Automattic/mongoose#6175]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
Schema is validated, but if schema has an array of schema's then it throws the below error:
Validation failed: undefined: Right-hand side of 'instanceof' is not callable, vertices: vertices is required.

**If the current behavior is a bug, please provide the steps to reproduce.**

Points.js
const Point = new mongoose.Schema({
            type: {type: String},
            latitude: {type: Number, required: [true, 'latitude is required. '], min: -90, max: 90},
            longitude: {type: Number, required: [true, 'longitude is required. '], min: -180, max: 180}
        });
module.exports = Point;

Polygon.js
const schema = new mongoose.Schema({
            guid: {type: String},
            name: {type: String, required: [true, 'name is required. ']},
            vertices: {type: [Point], required: [true, 'vertices is required. ']},
            creationTime: {type: Date, default: Date.now},
            lastUpdateTime: {type: Date, default: Date.now}
        });
module.exports = Polygon

Shapes.js
const polygon = {
            "type": "polygon",
            "name": "Test Polygon",
            "vertices": [
                {
                    "latitude": -37.81902680201739,
                    "longitude": 144.9821037054062
                },
                {
                    "latitude": -37.8196878692931,
                    "longitude": 144.98161554336548
                },
                {
                    "latitude": -37.82052267015026,
                    "longitude": 144.9816906452179
                },
                {
                    "latitude": -37.81902680201739,
                    "longitude": 144.9821037054062
                }
            ]
        };

Error: Right-hand side of 'instanceof' is not callable

**What is the expected behavior?**

**Please mention your node.js, mongoose and MongoDB version.**
node - 8.9.1
mongoose - 4.13.9

**Comment by lineus** [comment]
hi @RamAkunuru. I implemented your code like this:
```
#!/usr/bin/env node
'use strict'

const mongoose = require('../lib/test_db')
const Schema = mongoose.Schema
require('get-mongoose-versions').get()
  .then(versions => console.log(versions))

const Point = new Schema({
  type: { type: String },
  latitude: {
    type: Number,
    required: [true, 'latitude is required. '],
    min: -90,
    max: 90
  },
  longitude: {
    type: Number,
    required: [true, 'longitude is required. '],
    min: -180,
    max: 180
  }
})

const schema = new Schema({
  guid: { type: String },
  name: {
    type: String,
    required: [true, 'name is required. ']
  },
  vertices: {
    type: [Point],
    required: [true, 'vertices is required. ']
  },
  creationTime: { type: Date, default: Date.now },
  lastUpdateTime: { type: Date, default: Date.now }
})

const polygon = {
  'type': 'polygon',
  'name': 'Test Polygon',
  'vertices': [
    {
      'latitude': -37.81902680201739,
      'longitude': 144.9821037054062
    },
    {
      'latitude': -37.8196878692931,
      'longitude': 144.98161554336548
    },
    {
      'latitude': -37.82052267015026,
      'longitude': 144.9816906452179
    },
    {
      'latitude': -37.81902680201739,
      'longitude': 144.9821037054062
    }
  ]
}

const Poly = mongoose.model('poly', schema, 'poly')

const test = new Poly(polygon)

test.save().then((res) => {
  mongoose.connection.close()
  return console.log(res)
}, (rej) => {
  mongoose.connection.close()
  return console.error(rej)
})
```
here is the output:
```
InspiredMacPro:Help lineus$ ./mongoose4/6175/index.js
{ __v: 0,
  name: 'Test Polygon',
  _id: 5a95360fc96fe8f13627d073,
  lastUpdateTime: 2018-02-27T10:42:23.187Z,
  creationTime: 2018-02-27T10:42:23.187Z,
  vertices: 
   [ { latitude: -37.81902680201739,
       longitude: 144.9821037054062,
       _id: 5a95360fc96fe8f13627d077 },
     { latitude: -37.8196878692931,
       longitude: 144.98161554336548,
       _id: 5a95360fc96fe8f13627d076 },
     { latitude: -37.82052267015026,
       longitude: 144.9816906452179,
       _id: 5a95360fc96fe8f13627d075 },
     { latitude: -37.81902680201739,
       longitude: 144.9821037054062,
       _id: 5a95360fc96fe8f13627d074 } ] }
{ mongooseVersion: '4.13.9',
  mongodbNativeVersion: '2.2.34',
  nodeVersion: 'v8.9.1',
  mongodbServerVersion: '3.6.2' }
InspiredMacPro:Help lineus$ 
```
It looks like it's working for me. Is there something from my implementation of your schemas that differs from yours?

**Comment by lineus** [comment]
ok, sorry for the delay, I've never used mongoose in the browser before. here is my index.html file:
```
<html>
<head>
<script type = "text/javascript" src = "index.js"></script>
</head>
</html>
```
here is my main.js file # included in the index.html file after being browserified:
InspiredMacPro:6175 lineus$ browserify main.js > index.js
```
'use strict'

const mongoose = require('mongoose')
const Schema = mongoose.Schema

const Point = new Schema({
  latitude: {
    type: Number,
    required: [true, 'latitude is required. '],
    min: -90,
    max: 90
  },
  longitude: {
    type: Number,
    required: [true, 'longitude is required. '],
    min: -180,
    max: 180
  }
})

const schema = new Schema({
  name: {
    type: String,
    required: [true, 'name is required. ']
  },
  vertices: {
    type: [Point],
    required: [true, 'vertices is required. ']
  }
})

const polygon = {
  'name': 'Test Polygon',
  'vertices': [
    {
      'latitude': -37.81902680201739,
      'longitude': 144.9821037054062
    },
    {
      'latitude': -37.8196878692931,
      'longitude': 144.98161554336548
    },
    {
      'latitude': -37.82052267015026,
      'longitude': 144.9816906452179
    },
    {
      'latitude': -37.81902680201739,
      'longitude': 144.9821037054062
    }
  ]
}

const test = new mongoose.Document(polygon, schema)

test.validate(function (err) {
  if (err) { return console.error('err', err) }
  return console.log('yippie!')
})

```
here is the console output from firefox DE with mongoose 4.13.9:
```
message: "Validation failed: undefined: Document is not a function, vertices: vertices is required. "
```
I tried this code with other versions of mongoose, @4.10 - @4.11.8 all work as expected ( i get yippie! on console ), but starting at **@4.11.9** (tested sparsely, up to and including @4.13.11) i get the above error message.

**Comment by lineus** [comment]
looks like this is probably the same issue as  #5610. I believe the relevant change could have been e2790f6609d13206221a066932c34f682d4e3c0b.

**Comment by varunjayaraman** [comment]
Weird, I get a completely different error. @lineus i copied and pasted your script (with some minor modifications, i just browserified it into a file called `bundle.js` and changed the `src` in the html file) and instead I got:

```
bundle.js:66632 Uncaught TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at Document.$__getAllSubdocs (bundle.js:66632)
    at _getPathsToValidate (bundle.js:66059)
    at Document.$__validate (bundle.js:66141)
    at utils.promiseOrCallback.cb (bundle.js:66029)
    at Object.promiseOrCallback (bundle.js:77766)
    at Document.validate (bundle.js:66029)
    at Object.1.mongoose (bundle.js:55)
    at s (bundle.js:1)
    at e (bundle.js:1)
    at bundle.js:1
```

**Comment by vkarpov15** [comment]
Found the bug, fix will be in 5.0.12. @RamAkunuru thanks for your patience. Are you planning on upgrading to 5.x soon or do you want us to backport this fix to 4.x?

@lineus @varunjayaraman for the future, use the below script as an example for how to repro browser issues from Node, without using browserify. Should make it easier to reproduce bugs without having to go through a compile step.

```javascript
const mongoose = require('mongoose/browser');

const GITHUB_ISSUE = `gh-6175`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;

run().catch(error => console.error(error.stack));

async function run() {
  const { Schema } = mongoose;

  const Point = new Schema({
    latitude: {
      type: Number,
      required: [true, 'latitude is required. '],
      min: -90,
      max: 90
    },
    longitude: {
      type: Number,
      required: [true, 'longitude is required. '],
      min: -180,
      max: 180
    }
  })
  
  const schema = new Schema({
    name: {
      type: String,
      required: [true, 'name is required. ']
    },
    vertices: {
      type: [Point],
      required: [true, 'vertices is required. ']
    }
  })

  const polygon = {
    'name': 'Test Polygon',
    'vertices': [
      {
        'latitude': -37.81902680201739,
        'longitude': 144.9821037054062
      },
      {
        'latitude': -37.8196878692931,
        'longitude': 144.98161554336548
      },
      {
        'latitude': -37.82052267015026,
        'longitude': 144.9816906452179
      },
      {
        'latitude': -37.81902680201739,
        'longitude': 144.9821037054062
      }
    ]
  }

  const test = new mongoose.Document(polygon, schema)

  const err = await test.validate();
}
```
